sklearn/metrics/pairwise.py

```diff
@@ -256,6 +284,123 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
     return distances if squared else np.sqrt(distances, out=distances)
 
 
+def masked_euclidean_distances(X, Y=None, squared=False,
+                               missing_values="NaN", copy=True):
+    """Calculates euclidean distances in the presence of missing values
+
+    Considering the rows of X (and Y=X) as samples, compute the distance matrix
+    between each pair of samples. Similarly, if Y is not X, then compute the
+    distance matrix between each sample pair (i.e., each row pair) in X and Y.
+
+    When calculating the distance between a pair of samples, this formulation
```

Comment:
I would call this "average euclidean distance". I think I saw that term before and I think it's quite descriptive. Again, I find the description a bit overly complicated. Maybe something like "Distances are assumed zero if a value is missing, and are rescaled by the number of missing values." Maybe also mention that this kind of makes sense if the scale of all the values is the same? Basically we're assuming that on average the distance in each feature is the same... or maybe better that the average distance between the other features is a good proxy for the distance between the missing features?

Reply:
(I just realized that we're actually not computing average euclidean distance because we multiply by n_features in the end..)